File: core/src/main/java/com/toocol/ssh/core/shell/core/ActionLeftRightArrow.java
package com.toocol.ssh.core.shell.core;

import com.toocol.ssh.utilities.event.CharEvent;
import com.toocol.ssh.utilities.utils.CharUtil;

/**
 * @author ZhaoZhe (joezane.cn@gmail.com)
 * @date 2022/4/21 20:42
 */
public final class ActionLeftRightArrow extends ShellCharAction {
    @Override
    public CharEvent[] watch() {
        return new CharEvent[]{CharEvent.LEFT_ARROW, CharEvent.RIGHT_ARROW};
    }

    @Override
    public boolean act(Shell shell, CharEvent charEvent, char inChar) {
        if (shell.status.equals(Shell.Status.QUICK_SWITCH)) {
            return false;
        }
        int cursorX = shell.term.getCursorPosition()[0];
        if (inChar == CharUtil.LEFT_ARROW) {
            if (cursorX > shell.prompt.get().length()) {
                shell.term.cursorLeft();
            }
        } else {
            if (cursorX < (shell.currentPrint.length() + shell.prompt.get().length())) {
                shell.term.cursorRight();
            }
        }
        return false;
    }
}

File: core/src/main/java/com/toocol/termio/core/shell/core/ActionLeftRightArrow.kt
package com.toocol.termio.core.shell.core

import com.toocol.termio.utilities.event.CharEvent
import com.toocol.termio.utilities.utils.CharUtil

/**
 * @author ZhaoZhe (joezane.cn@gmail.com)
 * @date 2022/4/21 20:42
 */
class ActionLeftRightArrow : ShellCharAction() {
    override fun watch(): Array<CharEvent?> {
        return arrayOf(CharEvent.LEFT_ARROW, CharEvent.RIGHT_ARROW)
    }

    override fun actOn(shell: Shell, charEvent: CharEvent, inChar: Char): Boolean {
        if (shell.status == Shell.Status.QUICK_SWITCH) {
            return false
        }
        val cursorX = shell.term.cursorPosition[0]
        if (inChar == CharUtil.LEFT_ARROW) {
            if (cursorX > shell.prompt.get().length) {
                shell.term.cursorLeft()
            }
        } else {
            if (cursorX < shell.currentPrint.length + shell.prompt.get().length) {
                shell.term.cursorRight()
            }
        }
        return false
    }
}